code/modules/jobs/job_exp.dm

```diff
@@ -0,0 +1,325 @@
+		return_text += "<BR><BR>Jobs Not Unlocked:<UL><LI>"
+		return_text += jobs_locked.Join("</LI><LI>")
+		return_text += "</LI></UL>"
+	return return_text
+
+
+/client/proc/get_exp_living()
+	var/exp_living = text2num(prefs.exp[EXP_TYPE_LIVING])
+	return get_exp_format(exp_living)
+
+/proc/get_exp_format(expnum)
+	if(expnum > 60)
+		return num2text(round(expnum / 60)) + "h"
+	else if(expnum > 0)
+		return num2text(expnum) + "m"
+	else
+		return "0h"
+
+/proc/update_exp(mins, ann = FALSE)
+	if(!SSdbcore.Connect())
+		return -1
+	for(var/client/L in GLOB.clients)
+		if(L.is_afk())
+			continue
+		addtimer(CALLBACK(L,/client/proc/update_exp_list,mins,ann),10)
+		CHECK_TICK
+
+/proc/update_exp_db()
+	SSdbcore.MassInsert(format_table_name("role_time"),GLOB.exp_to_update,TRUE)
+	LAZYCLEARLIST(GLOB.exp_to_update)
+
+//Manual incrementing/updating
+/client/proc/update_exp_client(minutes, announce_changes = FALSE)
+	if(!src ||!ckey || !config.use_exp_tracking)
+		return
+	if(!SSdbcore.Connect())
+		return -1
+	var/datum/DBQuery/exp_read = SSdbcore.NewQuery("SELECT job, minutes FROM [format_table_name("role_time")] WHERE ckey = '[sanitizeSQL(ckey)]'")
+	if(!exp_read.Execute())
+		var/err = exp_read.ErrorMsg()
+		log_game("SQL ERROR during exp_update_client read. Error : \[[err]\]\n")
+		message_admins("SQL ERROR during exp_update_client read. Error : \[[err]\]\n")
+		return -1
+	var/list/play_records = list()
+	while(exp_read.NextRow())
+		play_records[exp_read.item[1]] = text2num(exp_read.item[2])
+
+	for(var/rtype in SSjob.name_occupations)
+		if(!play_records[rtype])
+			play_records[rtype] = 0
+	for(var/rtype in GLOB.exp_specialmap)
+		if(!play_records[rtype])
+			play_records[rtype] = 0
+	var/list/old_records = play_records.Copy()
+	if(mob.stat == CONSCIOUS && mob.mind.assigned_role)
+		play_records[EXP_TYPE_LIVING] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Living EXP!")
+		for(var/job in SSjob.name_occupations)
+			if(mob.mind.assigned_role == job)
+				play_records[job] += minutes
+				if(announce_changes)
+					to_chat(mob,"<span class='notice'>You got: [minutes] [job] EXP!")
+		if(mob.mind.special_role)
+			play_records[EXP_TYPE_SPECIAL] += minutes
+			if(announce_changes)
+				to_chat(mob,"<span class='notice'>You got: [minutes] Special EXP!")
+	else if(isobserver(mob))
+		play_records[EXP_TYPE_GHOST] += minutes
+		if(announce_changes)
+			to_chat(mob,"<span class='notice'>You got: [minutes] Ghost EXP!")
+	else if(minutes)	//Let "refresh" checks go through
+		return
+	prefs.exp = play_records
+
+	for(var/jtype in play_records)
+		if(play_records[jtype] != old_records[jtype])
+			var jobname = jtype
+			var time = play_records[jtype]
+			var/datum/DBQuery/update_query = SSdbcore.NewQuery("INSERT INTO [format_table_name("role_time")] (`ckey`, `job`, `minutes`) VALUES ('[sanitizeSQL(ckey)]', '[sanitizeSQL(jobname)]', '[sanitizeSQL(time)]') ON DUPLICATE KEY UPDATE minutes = VALUES(minutes)")
```

Comment:
It was supposed to be a manual proc call for debugging purposes. But since any db call gets auto denied from advanced proc call now, I should probably remove it.